lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs

```diff
@@ -8,5 +11,173 @@
+data ConfigParam = ConfigParam
+    deriving Show
+
+{-# ANN getBlockHeader ("HLint: ignore Use <$>" :: String) #-}
+getBlockHeader :: Get BlockHeader
+getBlockHeader =  (fromIntegral <$> getWord16be) >>= \s -> isolate s $ do
+    version <- getWord16be
+    sizeOfContent <- getWord32be
+
+    slotEpoch <- fromIntegral <$> getWord32be
+    slotId <- fromIntegral <$> getWord32be
+
+    chainLength <- getWord32be
+
+    contentHash <- getByteString 32 -- or 256 bits
+
+    parentHeaderHash <- getParentHeaderHash
+
+    -- TODO: Handle special case for BFT
+    -- TODO: Handle special case for Praos/Genesis
+
+    return $
+        BlockHeader
+            version
+            sizeOfContent
+            (SlotId slotId slotEpoch)
+            chainLength
+            contentHash
+            parentHeaderHash
+
+getBlock :: Get Block
+getBlock = do
+    header <- getBlockHeader
+    msgs <- isolate (fromIntegral $ contentSize header)
+        $ whileM (not <$> isEmpty) getBlockContent
+    return $ Block header msgs
+
+getBlockContent :: Get Message
+getBlockContent = do
+    size <- getWord16be
+    contentType <- fromIntegral <$> getWord8
+
+    let msgSize = fromIntegral size - 1
+
+    let unimpl = skip msgSize >> return (UnimplementedMessage contentType)
+
+    isolate msgSize $ case contentType of
+        0 -> unimpl
+        1 -> unimpl
+        2 -> unimpl
+        3 -> unimpl
+        4 -> unimpl
+        5 -> unimpl
+        other -> fail $ "Unexpected content type tag " ++ show other
+
+getParentHeaderHash :: Get (Maybe ByteString)
+getParentHeaderHash = getByteString 32 >>= \case
+    a | a == BS.pack (replicate 32 0)
+        -> return Nothing
+      | otherwise
+        -> return $ Just a
+
+
+{-------------------------------------------------------------------------------
+                            For development
+-------------------------------------------------------------------------------}
+
+dev :: IO ()
+dev = print $ runGet getBlock genesisBlock
+
+genesisBlock :: BL.ByteString
+genesisBlock = either error BL.fromStrict $ convertFromBase @ByteString Base16
+    "005200000000009f000000000000000000000000ffadebfecd59d9eaa12e903a\
+    \d58100f7c1e35899739c3d05d022835c069d2b4f000000000000000000000000\
+    \00000000000000000000000000000000000000000047000048000000005cc1c2\
+    \4900810200c200010108000000000000087001410f01840000000a01e030a694\
+    \b80dbba2d1b8a4b55652b03d96315c8414b054fa737445ac2d2a865c76002604\
+    \0001000000ff0005000006000000000000000000000000000000000000000000\
+    \0000000000002c020001833324c37869c122689a35917df53a4f2294a3a52f68\
+    \5e05f5f8e53b87e7ea452f000000000000000e"
+
+
+{-------------------------------------------------------------------------------
+                              Helpers
+-------------------------------------------------------------------------------}
+
+whileM :: Monad m => m Bool -> m a -> m [a]
+whileM p f = go
+  where
+    go = do
+        x <- p
+        if x then do
```

Comment:
`x` is rather confusing as a name here. Instead `predicate` or `cond` would show that it's actually not the same kind of data as `x'`.

Reply:
went with
```haskell
whileM :: Monad m => m Bool -> m a -> m [a]
whileM cond next = go
  where
    go = do
        c <- cond
        if c then do
            a <- next
            as <- go
            return (a : as)
        else return []
```